### ISSUE TITLE
Frivolous HUD suggestion

### ISSUE THREAD

**Issue by Chonkblonk** [fabiangreffrath/woof#784]
Completely unnecessary suggestion, but I'll just make it for the hell of it.

Do you think it would be neat to add support for the 3DO/Jaguar/PSX feature where on the player's XDeath, and if the relevant frames are present in the .wad, the face gib animation is played? Like as one of those optional features.

The original animation is played at like 10 tics per frame, so it doesn't look the liveliest, but if it starts on an OUCH face for maybe 20 tics and then runs through the animation at 2 tics per frame, it looks pretty energetic and lively.

Either way, just a goofy little gimmick thing which maybe most people wouldn't use, but I think it'd be a real neat touch.

**Comment by fabiangreffrath** [comment]
How are these frames called? Do you know of any PWAD that includes them? Where can I find more information?

Just to make one thing clear from the start, we will not include additional graphics in the engine.

**Comment by Chonkblonk** [comment]
The frames would be named:
**STFXDTH0**
**STFXDTH1**
**STFXDTH2**
**STFXDTH3**
**STFXDTH4**
**STFXDTH5**
Here's how they would look like (rip may not have been 100% perfect): 
**[https://i.imgur.com/qo1obbL.png](https://i.imgur.com/qo1obbL.png)**
The animation was first featured in the port for the Atari Jaguar, as well as the port for the 3DO. Most people have seen it in the Playstation ports.
**[https://www.youtube.com/watch?v=v4Sjivx4EDs](https://www.youtube.com/watch?v=v4Sjivx4EDs)**

The original animation plays at 10 tics per frame, disappearing after the last frame.
I personally think that it looks better starting at STOUCH2 for 10 tics, then going from STXDTH0 to STXDTH5 at 2 tics per frame, stopping at the last frame and remaining there.
That's how I implemented it in a current WIP project as a bonus patch one can use with GzDoom, if you want to see an example:
**[https://mega.nz/folder/z9ZVEAJJ#t4IEL6VGFY7tBBMIEOKXUA](https://mega.nz/folder/z9ZVEAJJ#t4IEL6VGFY7tBBMIEOKXUA)**

If you're legit interested in implementing this as a feature, I would suggest being able to set it at slow, where it plays at 10 tics per frame (your choice if you want it to disappear or not, but I think it looks best if the last frame stays), and then a fast where it plays faster similar to my example.
Obviously, no frames would be provided by Woof itself, and the behavior would default to the original Vanilla behavior (just the regular STFDEAD0), unless a loaded .wad provided their own frames, be that ones like the ripped ones, or ones of their own.

Again, this is totally unnecessary, but I also think it would be a cool feature.

**Comment by fabiangreffrath** [comment]
Okay, I think we could somehow bind the 6 graphics to the 9 S_PLAY_XDIE states, so we wouldn't even need a separate timer for this.

**Comment by fabiangreffrath** [comment]
There are 5 tics between each state in the S_PLAY_XDIE1 to S_PLAY_XDIE9 sequence.

https://github.com/fabiangreffrath/woof/blob/c1a7a8e71dd33f867b8aa6df8cb0b9f543878445/src/info.c#L290-L298

I think we could arrange something like "play STFXDTH0 for the first three states and then a new frame for each new state and keep the last frame for the last three states".

**Comment by fabiangreffrath** [comment]
> if you want to see an example: **https://mega.nz/folder/z9ZVEAJJ#t4IEL6VGFY7tBBMIEOKXUA**

Which of these iw the relevant WAD? Or, even better, could you please post a WAD with only the six relevant frames?

**Comment by Chonkblonk** [comment]
Oh, uh, 94BETA0.8.wad has the (modified) frames, 94SplatDefault has the code for implementing the animation for GzDoom's default HUD, while 94SplatNC-HUD.pk3 has the code for implementing the animation for the NC-HUD mod.
https://forum.zdoom.org/viewtopic.php?t=33079

Uh, anyway, I guess I could make a variant which is just the 'vanilla' (or close to vanilla) frames for GzDoom's default HUD.

**Comment by Chonkblonk** [comment]
Ok, so I added HEADSPLAT.wad to the Mega link, which has the code and frames for the default Doomguy face.
Since the source of the rips had a black background, I maybe had to improvise slightly when cutting them out, so they may not be 100% accurate to the original frames, but that's not really important because this is just an example and people will have to provide their own frames anyway (I could probably polish them up a little bit at some point as a community asset).

**Comment by fabiangreffrath** [comment]
Ad-hoc patch, seems to work quite well:
```patch
diff --git a/src/st_stuff.c b/src/st_stuff.c
index 4a8b2a3..69d8a5f 100644
--- a/src/st_stuff.c
+++ b/src/st_stuff.c
@@ -83,9 +83,10 @@ extern boolean inhelpscreens;
           (ST_NUMSTRAIGHTFACES+ST_NUMTURNFACES+ST_NUMSPECIALFACES)
 
 #define ST_NUMEXTRAFACES        2
+#define ST_NUMXDTHFACES         6
 
 #define ST_NUMFACES \
-          (ST_FACESTRIDE*ST_NUMPAINFACES+ST_NUMEXTRAFACES)
+          (ST_FACESTRIDE*ST_NUMPAINFACES+ST_NUMEXTRAFACES+ST_NUMXDTHFACES)
 
 #define ST_TURNOFFSET           (ST_NUMSTRAIGHTFACES)
 #define ST_OUCHOFFSET           (ST_TURNOFFSET + ST_NUMTURNFACES)
@@ -93,6 +94,7 @@ extern boolean inhelpscreens;
 #define ST_RAMPAGEOFFSET        (ST_EVILGRINOFFSET + 1)
 #define ST_GODFACE              (ST_NUMPAINFACES*ST_FACESTRIDE)
 #define ST_DEADFACE             (ST_GODFACE+1)
+#define ST_XDTHFACE             (ST_DEADFACE+1)
 
 #define ST_FACESX               143
 #define ST_FACESY               168
@@ -102,6 +104,7 @@ extern boolean inhelpscreens;
 #define ST_TURNCOUNT            (1*TICRATE)
 #define ST_OUCHCOUNT            (1*TICRATE)
 #define ST_RAMPAGEDELAY         (2*TICRATE)
+#define ST_XDTHCOUNT            10
 
 #define ST_MUCHPAIN             20
 
@@ -254,6 +257,7 @@ static patch_t *keys[NUMCARDS+3];
 
 // face status patches
 static patch_t *faces[ST_NUMFACES];
+static boolean have_xdthfaces;
 
 // face background
 static patch_t *faceback[MAXPLAYERS]; // killough 3/7/98: make array
@@ -520,6 +524,37 @@ int ST_calcPainOffset(void)
 //  dead > evil grin > turned head > straight ahead
 //
 
+static int ST_DeadFace(void)
+{
+  const int state = (plyr->mo->state - states) - mobjinfo[plyr->mo->type].xdeathstate;
+
+  if (have_xdthfaces && state >= 0)
+  {
+    static int count, index = -1;
+
+    if (state == 0 && index != 0)
+    {
+      count = ST_XDTHCOUNT;
+      index = 0;
+    }
+
+    if (count == 0)
+    {
+      count = ST_XDTHCOUNT;
+      index++;
+
+      if (index == ST_NUMXDTHFACES)
+        index = ST_NUMXDTHFACES - 1;
+    }
+
+    count--;
+
+    return ST_XDTHFACE + index;
+  }
+
+  return ST_DEADFACE;
+}
+
 void ST_updateFaceWidget(void)
 {
   int         i;
@@ -535,7 +570,7 @@ void ST_updateFaceWidget(void)
       if (!plyr->health)
         {
           priority = 9;
-          st_faceindex = ST_DEADFACE;
+          st_faceindex = ST_DeadFace();
           st_facecount = 1;
         }
     }
@@ -1044,6 +1079,17 @@ void ST_loadGraphics(void)
     }
   faces[facenum++] = W_CacheLumpName("STFGOD0", PU_STATIC);
   faces[facenum++] = W_CacheLumpName("STFDEAD0", PU_STATIC);
+
+  for (i = 0; i < ST_NUMXDTHFACES; i++)
+  {
+    sprintf(namebuf, "STFXDTH%d0", i);
+
+    if (W_CheckNumForName(namebuf) != -1)
+      faces[facenum++] = W_CacheLumpName(namebuf, PU_STATIC);
+    else
+      break;
+  }
+  have_xdthfaces = (i == ST_NUMXDTHFACES);
 }
 
 void ST_loadData(void)
@@ -1084,7 +1130,8 @@ void ST_unloadGraphics(void)
     Z_ChangeTag(faceback[i], PU_CACHE);
 
   for (i=0;i<ST_NUMFACES;i++)
-    Z_ChangeTag(faces[i], PU_CACHE);
+    if (faces[i])
+      Z_ChangeTag(faces[i], PU_CACHE);
 
   // Note: nobody ain't seen no unloading of stminus yet. Dude.
 }
```

**Comment by fabiangreffrath** [comment]
PS: Seems that there are different animations available for 3DO/Jaguar/PSX. Which one are these based on and how many frames do the others have?

**Comment by Chonkblonk** [comment]
Oh, uh, the ones I used are from the Jaguar port. I don't know if the ones from the 3DO port are actually different?
Looking closer at the ones from the Playstation version (which appear to be five frames), they are a little different, possibly they were redrawn to look better at the smaller scale.

I cannot find a site which has the frames from the 3DO version, and I can't actually find somewhere hosting the Jaguar ones either, possibly the frames I got were taken from a .gif showcasing the animation which has a black background. Apparently Kaiser may have them.

**Comment by fabiangreffrath** [comment]
It would be trivial to extend this to support 5 *or* 6 frames. But I am still not sure if its usefulness exceeds its maintenance burden... @rfomin ?

**Comment by rfomin** [comment]
I don't have an opinion on this. I saw that Inter Doom has some extended frames, maybe @JNechaevsky can give some advice.

**Comment by JNechaevsky** [comment]
Uh... I went even farther in ID and done:
* Doubled faces array to make player grinning/rampage with blood while invulnerability effect.
* Face explosion (Jaguar, gfx/behaviour).
* Face squishing while death by crushed (PSX, just gfx).

It is optional feature and strictly off if loaded PWAD contains replaced mugshot, but like many others, it was my personal fad, never considered to become a cross-port standard. 

I honestly don't know what to suggest. If needed, I have most of all console IWADs for inspection, all readable by Slade.

**Comment by fabiangreffrath** [comment]
Thanks @JNechaevsky! I see you bind the six "exploded face" frames to the first six states of the player's xdeathstate. This will result in 5 tics between each frame, which I found a bit fast, but I like how clean and straightforward that is. 🤔

**Comment by JNechaevsky** [comment]
> which I found a bit fast

True, actually, but there is more notable problem: to have an "explosion face" and go to xdeath state, player must receive very high amount of damage, and thus, screen goes to very intensive red for certain time. In other words, this animation is barely noticeable beneath pain palette, and palette fading normal takes far more time than explosion animation. 

AFAIR, Jaguar Doom's status bar was not affected by palette changing at all.

**Comment by fabiangreffrath** [comment]
Yes, right. I had to disable palette changes in order to debug this. I more and more doubt how useful this will be in real life...

**Comment by JNechaevsky** [comment]
🙃 Caconical names are "splat" ([screenshot](https://user-images.githubusercontent.com/21193394/199720979-4843e3d5-b909-4cfa-ad3b-d2b1a80cfc70.png)): 
> STSPLAT0
> STSPLAT1
> STSPLAT2
> STSPLAT3
> STSPLAT4
> STSPLAT5

**Comment by fabiangreffrath** [comment]
oh my...

**Comment by fabiangreffrath** [comment]
Ok, just did a bit of research: The sprites are called `STSPLAT0` to `STSPLAT5` in Kaiser's conres.zip/ConsoleDoom_Resources/Jaguar collection. The ConsoleDoom_Resources/Psx directory only contains a combined picture of all status bar graphics.

However, in PSXDoom ports, the lumps are also called like that, though the list ends at `STSPLAT4`:
https://github.com/BodbDearg/PsyDoom/blob/2816127288e0d3b89207825f8495f09c217c2fc3/game/Doom/UI/st_main.cpp#L85-L89
https://github.com/Erick194/PSXDOOM-RE/blob/2f15305a2030fbd0772c27a3fe811d32d94244cc/PSXDOOM/st_main.c#L69-L73

So, what do we know about Doom on 3DO?

**Comment by rfomin** [comment]
> So, what do we know about Doom on 3DO?

Doom on 3DO is based on the Jaguar version (like most Doom console ports from the 90s). I think only Jaguar and PSX versions are interesting.

**Comment by JNechaevsky** [comment]
Err... 3DO is kinda hard to read: https://github.com/Olde-Skuul/doom3do/blob/master/source/stmain.c (`FIRSTSPLAT`, `gibframe`?)

**Comment by rfomin** [comment]
> Err... 3DO is kinda hard to read

The source code is different, but the assets should be the same as the Jaguar version if I remember correctly from the Digital Foundry video: https://www.youtube.com/watch?v=784MUbDoLjQ

**Comment by fabiangreffrath** [comment]
> Doom on 3DO is based on the Jaguar version (like most Doom console ports from the 90s). I think only Jaguar and PSX versions are interesting.

Yes, probably. And they both already use the same name scheme. @Chonkblonk would you be okay to adapt to the canonical name scheme?

> Err... 3DO is kinda hard to read: https://github.com/Olde-Skuul/doom3do/blob/master/source/stmain.c (`FIRSTSPLAT`, `gibframe`?)

Seem to read in binary from the image file:

https://github.com/Olde-Skuul/doom3do/blob/f3ea056bc9659bc0e75efb928a406905fd918e10/source/stmain.c#L120

**Comment by JNechaevsky** [comment]
From what has been found in PSX Doom IWAD. It's a unified sheet of graphics:
<details>

<img width="782" alt="image" src="https://user-images.githubusercontent.com/21193394/199758551-ad87f98b-6b22-409b-ba37-d88938e8e8d0.png">

</details>

P.S. Link for archives for inspection: https://disk.yandex.ru/d/ls2DXfFFhNYmeQ

**Comment by Chonkblonk** [comment]
Oh wow, you guys did a lot while I was gone.

> _In other words, this animation is barely noticeable beneath pain palette, and palette fading normal takes far more time than explosion animation._
Well, that's why for my own implementation, I have the animation first linger on OUCH2 for 10 tics, because when you're getting killed you're already focused on the action, but if there's a brief pause like that first, the player would look down towards the status bar to see what's up, and _then_ see the gib animation.

I don't know if 10 tics would be enough to linger for really high damage though, and much longer than that would probably be too long. Potentially if the face graphics, or at the very least the SPLAT frames are unaffected by the palette change, but for the former I'd be worried that people would really dislike that, and I wouldn't want something like that foisted on people which they can't disable, and it also seems like it could be bothersome to implement, and maintain.

I don't know much about programming, but I wouldn't a fun little bonus feature like this to be a lot of trouble.

> _would you be okay to adapt to the canonical name scheme?_
Absolutely, it takes only a minute to change.

**Comment by Chonkblonk** [comment]
> _would you be okay to adapt to the canonical name scheme?_

Uh, it turns out that I cannot on my end, not exactly anyway. I fiddled with this, and GzDoom expects the status bar face frames to be in the STF namespace, so if I name the frames STSPLAT, the game won't recognize them, the closest they could be is STFPLAT.

This is NOT a big deal for ME though, as I could name them STSPLAT for the base resource, if Woof were to pick up on them, and then include duplicates of the frames named STFXDTH for GzDoom to detect. Very slightly wasteful, but only by a little bit.

**Comment by fabiangreffrath** [comment]
Does GZDoom *expect* the lumps to be called like this? Or do you have to tell it the lump names, anyway, but can't if they don't carry the `STF` prefix?

**Comment by Chonkblonk** [comment]
There could be another way, I'm not well versed in ZScript and SBARINFO coding, but as far as I can tell it wouldn't work with the setup I got help with creating.

`mugshot "XDeath"
{
    OUCH2 10;
    PLAT0 2; //not recognized
    SPLAT1 2; //too long
    XDTH2 2;
    XDTH3 2;
    XDTH4 2;
    XDTH5 -1;
}`

[https://i.imgur.com/VrMJRGW.png](https://i.imgur.com/VrMJRGW.png)

Again, it wouldn't be a big deal for me personally to have duplicate frames, and it probably wouldn't be that big of a deal for the few other people who wanted to make use of this feature. However, given that nobody is loading the Jaguar or 3DO .wads for source ports, is it that big of a deal to retain the STSPLAT name? STFXDTH would be a bit more consistent with all the other face frames which start with STF.

**Comment by fabiangreffrath** [comment]
It's a dilemma. On the one hand, I wouldn't want to introduce a feature that's canonical to other Doom implementations but break their naming scheme. On the other hand I wouldn't want to enforce a naming scheme that's contradictory to the expectations of a different source port that already supports this feature.

**Comment by Chonkblonk** [comment]
Well, STFXDTH would be more in line with the vanilla naming scheme than STSPLAT, the latter is an outlier. There's a lot about these ports which isn't in line with what was done on PC (or in most source ports).

**Comment by rfomin** [comment]
I think we should use the GZDoom scheme, I didn't know it supported it. It is much more practical.

**Comment by fabiangreffrath** [comment]
Alright, back to the old naming scheme then.

While at it: If we do not adhere to the canonical naming scheme anyway, we don't have to adhere to the "5 or 6 frames make a sequence" rule as well. So, we could support up to 9 frames, one for each state in the xdeathstate sequence. And we could also even support only 1 frame, the regular dead face is also only one frame after all.

**Comment by Chonkblonk** [comment]
> Alright, back to the old naming scheme then.
> 
> While at it: If we do not adhere to the canonical naming scheme anyway, we don't have to adhere to the "5 or 6 frames make a sequence" rule as well. So, we could support up to 9 frames, one for each state in the xdeathstate sequence. And we could also even support only 1 frame, the regular dead face is also only one frame after all.

Oh, that would actually be pretty cool, then someone could make an extended animation, or just a single gibbed face. It's really cool that you implemented this!

**Comment by Chonkblonk** [comment]
So, uh, since I'm not at all tech savvy and don't know how to actually _compile_ code into executables, I want to ask, how does it handle the red fading? Does it fade red just like always, or does it do something different so as to not be obscured if the player takes a super strong killing hit which bathes the screen red for a longer time?

**Comment by fabiangreffrath** [comment]
> So, uh, since I'm not at all tech savvy and don't know how to actually _compile_ code into executables, I want to ask, how does it handle the red fading? Does it fade red just like always, or does it do something different so as to not be obscured if the player takes a super strong killing hit which bathes the screen red for a longer time?

This commit doesn't affect palette changes at all.
